### PR TITLE
[ESIMD] Fix invoke_simd calls case with pointer passed to it

### PR DESCRIPTION
**Pull request by v-klochkov**
The helper function created during translation of invoke_simd must
accept a pointer to a function, not a reference to a pointer to a function.
That additional level of indirection is automatically resolved by compiler
for invoke_simd, but needs to be manually resolved/adjusted for the helper
function.

Signed-off-by: Vyacheslav N Klochkov <vyacheslav.n.klochkov@intel.com>

### CODE REVIEW

File: llvm/test/SYCLLowerIR/ESIMD/lower_invoke_simd.ll [hunk 1]
```diff
@@ -56,8 +59,18 @@ define dso_local spir_func noundef float @SPMD_CALLER(float noundef %x, <16 x fl
   %res3 = call spir_func float @_Z33__regcall3____builtin_invoke_simdXX(<16 x float> (<16 x float> (<16 x float>)*, <16 x float>)* @SIMD_CALL_HELPER, <16 x float> (<16 x float>)* %raw_fptr,  float %x)
 ; CHECK: %{{.*}} = call spir_func float @_Z33__regcall3____builtin_invoke_simdXX(<16 x float> (<16 x float> (<16 x float>)*, <16 x float>)* @SIMD_CALL_HELPER, <16 x float> (<16 x float>)* %{{.*}}, float %{{.*}})
 
-  %res4 = fadd float %res1, %res2
-  %res = fadd float %res3, %res4
+;---- The invoke_simd calls.
+; Test case when function pointer is passed the __builtin_invoke_simd
+; and the actual function can be deduced.
+  %res4 = call spir_func float @_Z33__regcall3____builtin_invoke_simdYY(<16 x float> (<16 x float> (<16 x float>)* addrspace(4)*, <16 x float>)* noundef nonnull @SIMD_CALL_HELPER_YY, <16 x float> (<16 x float>)* noundef nonnull @SIMD_CALLEE_YY, float noundef 0.000000e+00)
```
Comment (kbobrovs) [inline llvm/test/SYCLLowerIR/ESIMD/lower_invoke_simd.ll:65]: I think the test case has a bug. The second actual argument of the `invoke_simd` - `@SIMD_CALLEE_YY` - is a function pointer. It becomes the first formal argument of `@SIMD_CALL_HELPER_YY` - `<16 x float> (<16 x float>)* addrspace(4)* noundef align 8 dereferenceable(8) %f`, which is a pointer to function pointer.

Comment (v-klochkov) [reply llvm/test/SYCLLowerIR/ESIMD/lower_invoke_simd.ll:65]: That LLVM IR came from this test case below. There may be some error in IR generation, I am re-checking that again,
but do you agree that the test itself is correct?

```#include <CL/sycl.hpp>
#include <sycl/ext/intel/esimd.hpp>
#include <sycl/ext/oneapi/experimental/invoke_simd.hpp>
using namespace sycl::ext::oneapi::experimental;
namespace esimd = sycl::ext::intel::esimd;
constexpr int VL = 16;
typedef simd<float, VL> __regcall (*FuncType)(simd<float, VL>);

[[intel::device_indirectly_callable]] SYCL_EXTERNAL simd<float, VL> __regcall intersectFunc(simd<float, VL> arg) SYCL_ESIMD_FUNCTION;// {return arg;}

SYCL_EXTERNAL void testfn() {
    auto sg = this_sub_group();
    float test = 0;
    FuncType func = intersectFunc;
    float res = invoke_simd(sg, func, test);
}
```

Comment (v-klochkov) [reply llvm/test/SYCLLowerIR/ESIMD/lower_invoke_simd.ll:65]: 'pointer to function pointer' came from this 'else' branch here (see that a pointer to 'f', which is a pointer to a function is passed):

```
  if constexpr (is_function) {
    return __builtin_invoke_simd<true /*function*/, RetSpmd>(
        detail::simd_func_call_helper<N, Callable, T...>, f,
        detail::unwrap_uniform<T>::impl(args)...);
  } else {
    // TODO support functors and lambdas which are handled in this branch.
    // The limiting factor for now is that the LLVMIR data flow analysis
    // implemented in LowerInvokeSimd.cpp which, finds actual invoke_simd
    // target function, can't handle this case yet.
    return __builtin_invoke_simd<false /*functor/lambda*/, RetSpmd>(
        detail::simd_obj_call_helper<N, Callable, T...>, &f,
        detail::unwrap_uniform<T>::impl(args)...);
  }
```

File: sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp [hunk 2]
```diff
@@ -40,7 +40,7 @@
 template <bool IsFunc, class SpmdRet, class SimdCallee, class... SpmdArgs,
           class = std::enable_if_t<!IsFunc>>
 SYCL_EXTERNAL __regcall SpmdRet
-__builtin_invoke_simd(SimdCallee target, const void *obj, SpmdArgs... args)
+__builtin_invoke_simd(SimdCallee &&target, const void *obj, SpmdArgs... args)
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp:43]: As we talked, couple of comments:
- target is really not the user function, but the helper
- no need in `&&` as we know we pass by value always (from invoke_simd).
So, I suggest:
```cpp
template <bool IsFunc, class SpmdRet, class HelperFunc, class... UserSimdFuncAndSpmdArgs,
          class = std::enable_if_t<!IsFunc>>
SYCL_EXTERNAL __regcall SpmdRet
__builtin_invoke_simd(HelperFunc target, const void *obj, UserSimdFuncAndSpmdArgs... args)
```

File: sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp [hunk 3]
```diff
@@ -53,7 +53,7 @@ __builtin_invoke_simd(SimdCallee target, const void *obj, SpmdArgs... args)
 
 template <bool IsFunc, class SpmdRet, class SimdCallee, class... SpmdArgs,
           class = std::enable_if_t<IsFunc>>
-SYCL_EXTERNAL __regcall SpmdRet __builtin_invoke_simd(SimdCallee target,
+SYCL_EXTERNAL __regcall SpmdRet __builtin_invoke_simd(SimdCallee &&target,
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp:56]: Similarly here:
```
template <bool IsFunc, class SpmdRet, class HelperFunc, class... UserSimdFuncAndSpmdArgs,
          class = std::enable_if_t<IsFunc>>
SYCL_EXTERNAL __regcall SpmdRet __builtin_invoke_simd(HelperFunc helper, UserSimdFuncAndSpmdArgs... args)
```

File: sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp [hunk 4]
```diff
@@ -288,6 +290,27 @@ static constexpr bool is_function_ptr_or_ref_v =
...
     ;
 
+/// The variables typed as pointer to a function become lvalue-reference
+/// when passed to invoke_simd as universal pointer. That creates an
+/// additional indirection, which is resolved automatically by the compiler
+/// for the caller side of __builtin_invoke_simd, but which must be resolved
+/// manually during the creation of invoke-simd-helper.
+/// The class remove_ref_if_ref_to_pointer_callable helps to removes that
+/// unwanted indirection for variable-pointers to functions passed to
+/// invoke_simd.
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp:300]: I suggest to move this comment to the "call site", and leave here just a short description like "remove reference from 'reference to function pointer' type".

File: sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp [hunk 5]
```diff
@@ -288,6 +290,27 @@ static constexpr bool is_function_ptr_or_ref_v =
...
+template <typename Callable> struct remove_ref_if_ref_to_pointer_callable {
```
Comment (kbobrovs) [inline sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp:301]: This works only with function pointers, so I suggest
```suggestion
template <typename Callable> struct remove_ref_from_func_ptr_ref_type {
```

File: llvm/lib/SYCLLowerIR/LowerInvokeSimd.cpp [hunk 6]
```diff
@@ -250,14 +250,15 @@ bool collectUsesLookTrhoughMemAndCasts(Value *V,
...
+
     auto *St = dyn_cast<StoreInst>(UU);
 
     if (!St) {
       Uses.insert(U);
       continue;
     }
     // Current user is a store (of V) instruction, see if...
-    assert((V = St->getValueOperand()) &&
+    assert((V == St->getValueOperand()) &&
```
Comment (kbobrovs) [inline llvm/lib/SYCLLowerIR/LowerInvokeSimd.cpp:261]: Good catch!

**Comment by v-klochkov** [comment]
@kbobrovs - Thank you for the quick review. I made changes for your comments. Please re-review.
Also, I updated E2E test to verify the patch: https://github.com/intel/llvm-test-suite/pull/1238

**Review by kbobrovs — APPROVED** [review]

**Review by againull — DISMISSED** [review]

File: sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp [hunk 7]
```diff
@@ -37,10 +37,11 @@
 /// @tparam SimdCallee the type of the SIMD callee function (the "target"). Must
```
Comment (asudarsa) [inline sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp:37]: I think this comment needs rewording. Thanks

**Review by asudarsa — CHANGES_REQUESTED** [review]
Just a comment update requested. Otherwise looks ok. Thanks

**Review by kbobrovs — APPROVED** [review]

File: sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp (hunk 7 above)
Comment (v-klochkov) [reply sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp]: @asudarsa - thank you for the review.
Good catch. I updated the comment in this commit: 8c0c71b21a62f5399a3ed68090ea4523d02134be

**Review by againull — APPROVED** [review]

**Review by asudarsa — APPROVED** [review]
LGTM. Thanks

**Comment by v-klochkov** [comment]
Fail on CUDA and HIP is unrelated to this ESIMD/invoke_simd fix:
    Regression/fsycl-host-compiler.cpp

**Pull request merged**